### PR TITLE
Ignore nulls from terminal CTRL-space in LineEdit

### LINKED ISSUE #7001: REPL: ctrl-space should not be added to input
I'm finding that when I type a ctrl-space, it is being inserted into my repl input, causing the parser to get confused:

``` julia
julia> x = <ctrl-space>1<return>
<return>
<return>
ERROR: syntax: incomplete: premature end of input
```

This is on a day-old julia:

``` julia
julia> versioninfo()
Julia Version 0.3.0-prerelease+3217
Commit 5fcfb13* (2014-05-26 13:43 UTC)
Platform Info:
  System: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7-4500U CPU @ 1.80GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY)
  LAPACK: libopenblas
  LIBM: libopenlibm
```

### PR DESCRIPTION
**Pull request by c42f**
By default some terminals (eg, xterm) generate '\0' when CTRL-space is
pressed.  This is an invisible character, so hard to notice and delete,
but it breaks the input causing a parser error which persists into the
history.

AFAICT we must deal with this as early as possible by discarding any '\0'
characters immediately, as '\0' is used as a special value in the
rest of the keymap system.

Fixes #7001

### CODE REVIEW

File: base/LineEdit.jl [hunk 1]
```diff
@@ -725,6 +725,8 @@ function match_input(k::Dict, s, term=terminal(s), cs=Char[], keymap = k)
     # return an empty keymap function
     eof(term) && return keymap_fcn(nothing, "")
     c = read(term, Char)
+    # Ignore any '\0' (CTRL-space in xterm) - it's used by a special value elsewhere
```
Comment (tkelman) [inline base/LineEdit.jl]: as a special value?

**Comment by tkelman** [comment]
could this be added in the repl tests?

File: base/LineEdit.jl (hunk 1 above)
Comment (c42f) [inline base/LineEdit.jl]: Good question.  It seems to be used as a special placeholder for the wildcard match of any character - in the keybinding syntax `'*'` is transformed into `'\0'` when normalizing a key binding:

https://github.com/JuliaLang/julia/blob/master/base/LineEdit.jl#L644
https://github.com/JuliaLang/julia/blob/master/base/LineEdit.jl#L650

The main point being that the LineEdit code assumes it's not possible to generate a `'\0'` from the terminal, but when it is possible, things break.

Comment (c42f) [inline base/LineEdit.jl]: There, I've updated the commit in an attempt to explain this better.

**Comment by c42f** [comment]
@tkelman tests added.  The way I ended up having to do this is a bit brittle, but the fake repl doesn't display values correctly so I had to go for a negative test rather than a positive one.

File: test/repl.jl [hunk 2]
```diff
@@ -137,6 +145,7 @@ if !is_windows() || Sys.windows_version() >= Sys.WINDOWS_VISTA_VER
     # yield make sure this got processed
     readuntil(stdout_read, "1+1")
     close(t)
+    readuntil(stdout_read, "\n\n")
```
Comment (c42f) [inline test/repl.jl]: By the way, you can tell this line isn't strictly necessary for the test above, but I had to add it when the 7001 test was placed lower in the file: the reads are out of sync with the writes by this point in the integration test.

**Review by Keno — APPROVED** [review]
LGTM

**Comment by c42f** [comment]
Great to have this merged. Thanks for the reviews.

**Pull request merged**